lock-api/src/main/java/com/palantir/lock/client/TimestampCorroboratingTimelockService.java

```diff
@@ -127,26 +134,39 @@ public ConjureStartTransactionsResponse startTransactions(ConjureStartTransactio
             ToLongFunction<T> lowerBoundExtractor,
             ToLongFunction<T> upperBoundExtractor,
             OperationType operationType) {
+        // take snapshot before making the request
+        Instant wallClockTimeBeforeRequest = Instant.now();
         TimestampBounds timestampBounds = getTimestampBounds();
         T timestampContainer = timestampContainerSupplier.get();
 
         long lowerFreshTimestamp = lowerBoundExtractor.applyAsLong(timestampContainer);
         long upperFreshTimestamp = upperBoundExtractor.applyAsLong(timestampContainer);
+
         checkTimestamp(timestampBounds, operationType, lowerFreshTimestamp, upperFreshTimestamp);
-        updateLowerBound(operationType, upperFreshTimestamp);
+
+        TimestampBoundsRecord boundsRecord = ImmutableTimestampBoundsRecord.builder()
+                .operationType(operationType)
+                .inclusiveLowerBoundFromLastRequest(lowerFreshTimestamp)
+                .inclusiveUpperBoundFromLastRequest(upperFreshTimestamp)
+                .wallClockTimeBeforeRequest(wallClockTimeBeforeRequest)
+                .wallClockTimeAfterResponse(Instant.now())
+                .build();
+        updateLowerBound(operationType, boundsRecord);
         return timestampContainer;
     }
 
     @VisibleForTesting
     TimestampBounds getTimestampBounds() {
-        long threadLocalLowerBoundFromTimestamps = lowerBoundFromTimestamps.get();
-        long threadLocalLowerBoundFromTransactions = lowerBoundFromTransactions.get();
-        return ImmutableTimestampBounds.of(threadLocalLowerBoundFromTimestamps, threadLocalLowerBoundFromTransactions);
+        return ImmutableTimestampBounds.builder()
+                .boundFromFreshTimestamps(lowerBoundFromFreshTimestamps.get())
+                .boundFromCommitTimestamps(lowerBoundFromCommitTimestamps.get())
+                .boundFromTransactions(lowerBoundFromTransaction.get())
+                .build();
     }
 
     private void checkTimestamp(
             TimestampBounds bounds, OperationType type, long lowerFreshTimestamp, long upperFreshTimestamp) {
-        if (lowerFreshTimestamp <= Math.max(bounds.boundFromTimestamps(), bounds.boundFromTransactions())) {
+        if (lowerFreshTimestamp <= bounds.getMaxLowerBound()) {
             timestampViolationCallback.run();
             throw clocksWentBackwards(bounds, type, lowerFreshTimestamp, upperFreshTimestamp);
```

Comment:
We don't seem to be doing anything with the new information. Surely we should at least log the three bounds with timestamps, the timestamp of the comparison, maybe the current timestamp etc?

Reply:
Most of these was being done but added more logging for wall clock times of current request